### ISSUE TITLE
Permission errors when testing the package installed in the system path (e.g. by an admin)

### ISSUE THREAD

**Issue by avalentino** [fatiando/pooch#248]
Running the test suite when the package is installed in the system path (e.g. by an administrator) generates several errors. 

**Full code that generated the error**

```
$ python3 -m pytest --pyargs pooch -v -k "not test_ftp_downloader and not test_downloader_progressbar_ftp"
```

**Full error message**

<details>

```
======================================================= test session starts ========================================================
platform linux -- Python 3.9.2, pytest-6.0.2, py-1.10.0, pluggy-0.13.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/antonio/debian/git/pooch
collected 82 items / 2 deselected / 80 selected                                                                                    

pooch/tests/test_core.py::test_retrieve PASSED                                                                               [  1%]
pooch/tests/test_core.py::test_retrieve_fname PASSED                                                                         [  2%]
pooch/tests/test_core.py::test_retrieve_default_path PASSED                                                                  [  3%]
pooch/tests/test_core.py::test_pooch_local FAILED                                                                            [  5%]
pooch/tests/test_core.py::test_pooch_custom_url PASSED                                                                       [  6%]
pooch/tests/test_core.py::test_pooch_download PASSED                                                                         [  7%]
pooch/tests/test_core.py::test_pooch_download_retry_off_by_default PASSED                                                    [  8%]
pooch/tests/test_core.py::test_pooch_download_retry PASSED                                                                   [ 10%]
pooch/tests/test_core.py::test_pooch_download_retry_fails_eventually PASSED                                                  [ 11%]
pooch/tests/test_core.py::test_pooch_logging_level PASSED                                                                    [ 12%]
pooch/tests/test_core.py::test_pooch_update PASSED                                                                           [ 13%]
pooch/tests/test_core.py::test_pooch_corrupted FAILED                                                                        [ 15%]
pooch/tests/test_core.py::test_pooch_file_not_in_registry PASSED                                                             [ 16%]
pooch/tests/test_core.py::test_pooch_load_registry PASSED                                                                    [ 17%]
pooch/tests/test_core.py::test_pooch_load_registry_comments PASSED                                                           [ 18%]
pooch/tests/test_core.py::test_pooch_load_registry_fileobj PASSED                                                            [ 20%]
pooch/tests/test_core.py::test_pooch_load_registry_custom_url PASSED                                                         [ 21%]
pooch/tests/test_core.py::test_pooch_load_registry_invalid_line PASSED                                                       [ 22%]
pooch/tests/test_core.py::test_check_availability PASSED                                                                     [ 23%]
pooch/tests/test_core.py::test_check_availability_on_ftp PASSED                                                              [ 25%]
pooch/tests/test_core.py::test_fetch_with_downloader PASSED                                                                  [ 26%]
pooch/tests/test_core.py::test_invalid_hash_alg FAILED                                                                       [ 27%]
pooch/tests/test_core.py::test_alternative_hashing_algorithms FAILED                                                         [ 28%]
pooch/tests/test_core.py::test_download_action PASSED                                                                        [ 30%]
pooch/tests/test_core.py::test_stream_download[tiny-data.txt] PASSED                                                         [ 31%]
pooch/tests/test_core.py::test_stream_download[subdir/tiny-data.txt] PASSED                                                  [ 32%]
pooch/tests/test_downloaders.py::test_unsupported_protocol PASSED                                                            [ 33%]
pooch/tests/test_downloaders.py::test_sftp_downloader PASSED                                                                 [ 35%]
pooch/tests/test_downloaders.py::test_sftp_downloader_fail_if_file_object PASSED                                             [ 36%]
pooch/tests/test_downloaders.py::test_sftp_downloader_fail_if_paramiko_missing SKIPPED                                       [ 37%]
pooch/tests/test_downloaders.py::test_downloader_progressbar_fails[HTTPDownloader] SKIPPED                                   [ 38%]
pooch/tests/test_downloaders.py::test_downloader_progressbar_fails[FTPDownloader] SKIPPED                                    [ 40%]
pooch/tests/test_downloaders.py::test_downloader_progressbar_fails[SFTPDownloader] SKIPPED                                   [ 41%]
pooch/tests/test_downloaders.py::test_downloader_progressbar PASSED                                                          [ 42%]
pooch/tests/test_downloaders.py::test_downloader_progressbar_sftp PASSED                                                     [ 43%]
pooch/tests/test_downloaders.py::test_downloader_arbitrary_progressbar PASSED                                                [ 45%]
pooch/tests/test_integration.py::test_create_and_fetch PASSED                                                                [ 46%]
pooch/tests/test_processors.py::test_decompress[auto] PASSED                                                                 [ 47%]
pooch/tests/test_processors.py::test_decompress[lzma] PASSED                                                                 [ 48%]
pooch/tests/test_processors.py::test_decompress[xz] PASSED                                                                   [ 50%]
pooch/tests/test_processors.py::test_decompress[bz2] PASSED                                                                  [ 51%]
pooch/tests/test_processors.py::test_decompress[gz] PASSED                                                                   [ 52%]
pooch/tests/test_processors.py::test_decompress[name] PASSED                                                                 [ 53%]
pooch/tests/test_processors.py::test_decompress_fails PASSED                                                                 [ 55%]
pooch/tests/test_processors.py::test_extractprocessor_fails PASSED                                                           [ 56%]
pooch/tests/test_processors.py::test_processors[onefile-Unzip-default_dir] PASSED                                            [ 57%]
pooch/tests/test_processors.py::test_processors[onefile-Unzip-custom_dir] PASSED                                             [ 58%]
pooch/tests/test_processors.py::test_processors[onefile-Untar-default_dir] PASSED                                            [ 60%]
pooch/tests/test_processors.py::test_processors[onefile-Untar-custom_dir] PASSED                                             [ 61%]
pooch/tests/test_processors.py::test_processors[all_files-Unzip-default_dir] PASSED                                          [ 62%]
pooch/tests/test_processors.py::test_processors[all_files-Unzip-custom_dir] PASSED                                           [ 63%]
pooch/tests/test_processors.py::test_processors[all_files-Untar-default_dir] PASSED                                          [ 65%]
pooch/tests/test_processors.py::test_processors[all_files-Untar-custom_dir] PASSED                                           [ 66%]
pooch/tests/test_processors.py::test_processors[nested-Unzip-default_dir] PASSED                                             [ 67%]
pooch/tests/test_processors.py::test_processors[nested-Unzip-custom_dir] PASSED                                              [ 68%]
pooch/tests/test_processors.py::test_processors[nested-Untar-default_dir] PASSED                                             [ 70%]
pooch/tests/test_processors.py::test_processors[nested-Untar-custom_dir] PASSED                                              [ 71%]
pooch/tests/test_utils.py::test_unique_name_long PASSED                                                                      [ 72%]
pooch/tests/test_utils.py::test_make_local_storage_parallel[threads] PASSED                                                  [ 73%]
pooch/tests/test_utils.py::test_make_local_storage_parallel[processes] PASSED                                                [ 75%]
pooch/tests/test_utils.py::test_local_storage_makedirs_permissionerror PASSED                                                [ 76%]
pooch/tests/test_utils.py::test_local_storage_newfile_permissionerror PASSED                                                 [ 77%]
pooch/tests/test_utils.py::test_registry_builder FAILED                                                                      [ 78%]
pooch/tests/test_utils.py::test_registry_builder_recursive FAILED                                                            [ 80%]
pooch/tests/test_utils.py::test_parse_url PASSED                                                                             [ 81%]
pooch/tests/test_utils.py::test_file_hash_invalid_algorithm PASSED                                                           [ 82%]
pooch/tests/test_utils.py::test_hash_matches[sha256] PASSED                                                                  [ 83%]
pooch/tests/test_utils.py::test_hash_matches[sha512] PASSED                                                                  [ 85%]
pooch/tests/test_utils.py::test_hash_matches[md5] PASSED                                                                     [ 86%]
pooch/tests/test_utils.py::test_hash_matches_strict[sha256] PASSED                                                           [ 87%]
pooch/tests/test_utils.py::test_hash_matches_strict[sha512] PASSED                                                           [ 88%]
pooch/tests/test_utils.py::test_hash_matches_strict[md5] PASSED                                                              [ 90%]
pooch/tests/test_utils.py::test_hash_matches_none PASSED                                                                     [ 91%]
pooch/tests/test_utils.py::test_hash_matches_uppercase[sha256] PASSED                                                        [ 92%]
pooch/tests/test_utils.py::test_hash_matches_uppercase[sha512] PASSED                                                        [ 93%]
pooch/tests/test_utils.py::test_hash_matches_uppercase[md5] PASSED                                                           [ 95%]
pooch/tests/test_utils.py::test_temporary_file PASSED                                                                        [ 96%]
pooch/tests/test_utils.py::test_temporary_file_path PASSED                                                                   [ 97%]
pooch/tests/test_utils.py::test_temporary_file_exception PASSED                                                              [ 98%]
pooch/tests/test_version.py::test_version PASSED                                                                             [100%]

============================================================= FAILURES =============================================================
_________________________________________________________ test_pooch_local _________________________________________________________

path = '/usr/lib/python3/dist-packages/pooch/tests/data', env = None

    def make_local_storage(path, env=None):
        """
        Create the local cache directory and make sure it's writable.
    
        Parameters
        ----------
        path : str or PathLike
            The path to the local data storage folder.
        env : str or None
            An environment variable that can be used to overwrite *path*. Only used
            in the error message in case the folder is not writable.
        """
        path = str(path)
        # Check that the data directory is writable
        try:
            if not os.path.exists(path):
                action = "create"
                # When running in parallel, it's possible that multiple jobs will
                # try to create the path at the same time. Use exist_ok to avoid
                # raising an error.
                os.makedirs(path, exist_ok=True)
            else:
                action = "write to"
>               with tempfile.NamedTemporaryFile(dir=path):

/usr/lib/python3/dist-packages/pooch/utils.py:285: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

mode = 'w+b', buffering = -1, encoding = None, newline = None, suffix = '', prefix = 'tmp'
dir = '/usr/lib/python3/dist-packages/pooch/tests/data', delete = True

    def NamedTemporaryFile(mode='w+b', buffering=-1, encoding=None,
                           newline=None, suffix=None, prefix=None,
                           dir=None, delete=True, *, errors=None):
        """Create and return a temporary file.
        Arguments:
        'prefix', 'suffix', 'dir' -- as for mkstemp.
        'mode' -- the mode argument to io.open (default "w+b").
        'buffering' -- the buffer size argument to io.open (default -1).
        'encoding' -- the encoding argument to io.open (default None)
        'newline' -- the newline argument to io.open (default None)
        'delete' -- whether the file is deleted on close (default True).
        'errors' -- the errors argument to io.open (default None)
        The file is created as mkstemp() would do it.
    
        Returns an object with a file-like interface; the name of the file
        is accessible as its 'name' attribute.  The file will be automatically
        deleted when it is closed unless the 'delete' argument is set to False.
        """
    
        prefix, suffix, dir, output_type = _sanitize_params(prefix, suffix, dir)
    
        flags = _bin_openflags
    
        # Setting O_TEMPORARY in the flags causes the OS to delete
        # the file when it is closed.  This is only supported by Windows.
        if _os.name == 'nt' and delete:
            flags |= _os.O_TEMPORARY
    
>       (fd, name) = _mkstemp_inner(dir, prefix, suffix, flags, output_type)

/usr/lib/python3.9/tempfile.py:680: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

dir = '/usr/lib/python3/dist-packages/pooch/tests/data', pre = 'tmp', suf = '', flags = 131266, output_type = <class 'str'>

    def _mkstemp_inner(dir, pre, suf, flags, output_type):
        """Code common to mkstemp, TemporaryFile, and NamedTemporaryFile."""
    
        names = _get_candidate_names()
        if output_type is bytes:
            names = map(_os.fsencode, names)
    
        for seq in range(TMP_MAX):
            name = next(names)
            file = _os.path.join(dir, pre + name + suf)
            _sys.audit("tempfile.mkstemp", file)
            try:
>               fd = _os.open(file, flags, 0o600)
E               PermissionError: [Errno 13] Permission denied: '/usr/lib/python3/dist-packages/pooch/tests/data/tmpu__upbsb'

/usr/lib/python3.9/tempfile.py:390: PermissionError

The above exception was the direct cause of the following exception:

    def test_pooch_local():
        "Setup a pooch that already has the local data and test the fetch."
        pup = Pooch(path=DATA_DIR, base_url="some bogus URL", registry=REGISTRY)
        true = os.path.join(DATA_DIR, "tiny-data.txt")
>       fname = pup.fetch("tiny-data.txt")

/usr/lib/python3/dist-packages/pooch/tests/test_core.py:113: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3/dist-packages/pooch/core.py:520: in fetch
    make_local_storage(str(self.abspath))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

path = '/usr/lib/python3/dist-packages/pooch/tests/data', env = None

    def make_local_storage(path, env=None):
        """
        Create the local cache directory and make sure it's writable.
    
        Parameters
        ----------
        path : str or PathLike
            The path to the local data storage folder.
        env : str or None
            An environment variable that can be used to overwrite *path*. Only used
            in the error message in case the folder is not writable.
        """
        path = str(path)
        # Check that the data directory is writable
        try:
            if not os.path.exists(path):
                action = "create"
                # When running in parallel, it's possible that multiple jobs will
                # try to create the path at the same time. Use exist_ok to avoid
                # raising an error.
                os.makedirs(path, exist_ok=True)
            else:
                action = "write to"
                with tempfile.NamedTemporaryFile(dir=path):
                    pass
        except PermissionError as error:
            message = [
                str(error),
                f"| Pooch could not {action} data cache folder '{path}'.",
                "Will not be able to download data files.",
            ]
            if env is not None:
                message.append(
                    f"Use environment variable '{env}' to specify a different location."
                )
>           raise PermissionError(" ".join(message)) from error
E           PermissionError: [Errno 13] Permission denied: '/usr/lib/python3/dist-packages/pooch/tests/data/tmpu__upbsb' | Pooch could not write to data cache folder '/usr/lib/python3/dist-packages/pooch/tests/data'. Will not be able to download data files.

/usr/lib/python3/dist-packages/pooch/utils.py:297: PermissionError


[CUT]


_________________________________________________ test_registry_builder_recursive __________________________________________________

path = '/usr/lib/python3/dist-packages/pooch/tests/data/store', env = None

    def make_local_storage(path, env=None):
        """
        Create the local cache directory and make sure it's writable.
    
        Parameters
        ----------
        path : str or PathLike
            The path to the local data storage folder.
        env : str or None
            An environment variable that can be used to overwrite *path*. Only used
            in the error message in case the folder is not writable.
        """
        path = str(path)
        # Check that the data directory is writable
        try:
            if not os.path.exists(path):
                action = "create"
                # When running in parallel, it's possible that multiple jobs will
                # try to create the path at the same time. Use exist_ok to avoid
                # raising an error.
                os.makedirs(path, exist_ok=True)
            else:
                action = "write to"
>               with tempfile.NamedTemporaryFile(dir=path):

/usr/lib/python3/dist-packages/pooch/utils.py:285: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

mode = 'w+b', buffering = -1, encoding = None, newline = None, suffix = '', prefix = 'tmp'
dir = '/usr/lib/python3/dist-packages/pooch/tests/data/store', delete = True

    def NamedTemporaryFile(mode='w+b', buffering=-1, encoding=None,
                           newline=None, suffix=None, prefix=None,
                           dir=None, delete=True, *, errors=None):
        """Create and return a temporary file.
        Arguments:
        'prefix', 'suffix', 'dir' -- as for mkstemp.
        'mode' -- the mode argument to io.open (default "w+b").
        'buffering' -- the buffer size argument to io.open (default -1).
        'encoding' -- the encoding argument to io.open (default None)
        'newline' -- the newline argument to io.open (default None)
        'delete' -- whether the file is deleted on close (default True).
        'errors' -- the errors argument to io.open (default None)
        The file is created as mkstemp() would do it.
    
        Returns an object with a file-like interface; the name of the file
        is accessible as its 'name' attribute.  The file will be automatically
        deleted when it is closed unless the 'delete' argument is set to False.
        """
    
        prefix, suffix, dir, output_type = _sanitize_params(prefix, suffix, dir)
    
        flags = _bin_openflags
    
        # Setting O_TEMPORARY in the flags causes the OS to delete
        # the file when it is closed.  This is only supported by Windows.
        if _os.name == 'nt' and delete:
            flags |= _os.O_TEMPORARY
    
>       (fd, name) = _mkstemp_inner(dir, prefix, suffix, flags, output_type)

/usr/lib/python3.9/tempfile.py:680: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

dir = '/usr/lib/python3/dist-packages/pooch/tests/data/store', pre = 'tmp', suf = '', flags = 131266, output_type = <class 'str'>

    def _mkstemp_inner(dir, pre, suf, flags, output_type):
        """Code common to mkstemp, TemporaryFile, and NamedTemporaryFile."""
    
        names = _get_candidate_names()
        if output_type is bytes:
            names = map(_os.fsencode, names)
    
        for seq in range(TMP_MAX):
            name = next(names)
            file = _os.path.join(dir, pre + name + suf)
            _sys.audit("tempfile.mkstemp", file)
            try:
>               fd = _os.open(file, flags, 0o600)
E               PermissionError: [Errno 13] Permission denied: '/usr/lib/python3/dist-packages/pooch/tests/data/store/tmps86xf6py'

/usr/lib/python3.9/tempfile.py:390: PermissionError

The above exception was the direct cause of the following exception:

    def test_registry_builder_recursive():
        "Check that the registry builder works in recursive mode"
        outfile = NamedTemporaryFile(delete=False)
        # Need to close the file before writing to it.
        outfile.close()
        try:
            make_registry(DATA_DIR, outfile.name, recursive=True)
            with open(outfile.name) as fout:
                registry = fout.read()
            assert registry == REGISTRY_RECURSIVE
            # Check that the registry can be used.
            pup = Pooch(path=DATA_DIR, base_url="some bogus URL", registry={})
            pup.load_registry(outfile.name)
>           assert os.path.join(DATA_DIR, "tiny-data.txt") == pup.fetch("tiny-data.txt")

/usr/lib/python3/dist-packages/pooch/tests/test_utils.py:174: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3/dist-packages/pooch/core.py:520: in fetch
    make_local_storage(str(self.abspath))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

path = '/usr/lib/python3/dist-packages/pooch/tests/data/store', env = None

    def make_local_storage(path, env=None):
        """
        Create the local cache directory and make sure it's writable.
    
        Parameters
        ----------
        path : str or PathLike
            The path to the local data storage folder.
        env : str or None
            An environment variable that can be used to overwrite *path*. Only used
            in the error message in case the folder is not writable.
        """
        path = str(path)
        # Check that the data directory is writable
        try:
            if not os.path.exists(path):
                action = "create"
                # When running in parallel, it's possible that multiple jobs will
                # try to create the path at the same time. Use exist_ok to avoid
                # raising an error.
                os.makedirs(path, exist_ok=True)
            else:
                action = "write to"
                with tempfile.NamedTemporaryFile(dir=path):
                    pass
        except PermissionError as error:
            message = [
                str(error),
                f"| Pooch could not {action} data cache folder '{path}'.",
                "Will not be able to download data files.",
            ]
            if env is not None:
                message.append(
                    f"Use environment variable '{env}' to specify a different location."
                )
>           raise PermissionError(" ".join(message)) from error
E           PermissionError: [Errno 13] Permission denied: '/usr/lib/python3/dist-packages/pooch/tests/data/store/tmps86xf6py' | Pooch could not write to data cache folder '/usr/lib/python3/dist-packages/pooch/tests/data/store'. Will not be able to download data files.

/usr/lib/python3/dist-packages/pooch/utils.py:297: PermissionError
========================================================= warnings summary =========================================================
/usr/lib/python3/dist-packages/invoke/loader.py:3
  /usr/lib/python3/dist-packages/invoke/loader.py:3: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
===================================================== short test summary info ======================================================
FAILED pooch/tests/test_core.py::test_pooch_local - PermissionError: [Errno 13] Permission denied: '/usr/lib/python3/dist-package...
FAILED pooch/tests/test_core.py::test_pooch_corrupted - PermissionError: [Errno 13] Permission denied: '/usr/lib/python3/dist-pac...
FAILED pooch/tests/test_core.py::test_invalid_hash_alg - PermissionError: [Errno 13] Permission denied: '/usr/lib/python3/dist-pa...
FAILED pooch/tests/test_core.py::test_alternative_hashing_algorithms - PermissionError: [Errno 13] Permission denied: '/usr/lib/p...
FAILED pooch/tests/test_utils.py::test_registry_builder - PermissionError: [Errno 13] Permission denied: '/usr/lib/python3/dist-p...
FAILED pooch/tests/test_utils.py::test_registry_builder_recursive - PermissionError: [Errno 13] Permission denied: '/usr/lib/pyth...
================================ 6 failed, 70 passed, 4 skipped, 2 deselected, 1 warning in 29.16s =================================
```

</details>

**System information**

* Operating system: GNU/Linux x86_64 (Debian sid)
* Python installation (Anaconda, system, ETS): system
* Version of Python: 3.9.2
* Version of this package: 1.4.0

**Comment by leouieda** [comment]
@avalentino thanks for reporting! This seems to be a problem with some our tests that try to write temporary files to the install path. Mostly they setup the Pooch data directory in `DATA_DIR` which is the test data folder in the install path. 

The fix would be to create a duplicate of this folder in a temporary directory for every test. That would fix the permission issues and is the right thing to do since it avoids accidentally writing to the install path.

I'm a bit short on time to implement this so if anyone is willing to take this on, that would be great.
